tests/unit/test_charm.py

```diff
@@ -0,0 +1,112 @@
+import unittest
+from unittest.mock import PropertyMock, patch
+
+from charm import SlurmdCharm
+from ops.model import ActiveStatus, BlockedStatus
+from ops.testing import Harness
+
+
+class TestCharm(unittest.TestCase):
+    def setUp(self) -> None:
+        self.harness = Harness(SlurmdCharm)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    @patch("ops.framework.EventBase.defer")
+    def test_check_etcd_fail(self, defer):
+        self.harness.charm.on.check_etcd.emit()
+        defer.assert_called()
+
+    @patch("charm.SlurmdCharm._on_slurmctld_started")
+    @patch("json.loads", lambda _: ["test"])
+    @patch("charm.SlurmdCharm.hostname", new_callable=PropertyMock(return_value="test"))
+    @patch("omnietcd3.Etcd3AuthClient.get")
+    @patch("charm.SlurmdCharm.etcd_ca_cert", new_callable=PropertyMock(return_value=""))
+    @patch("charm.SlurmdCharm.etcd_use_tls", new_callable=PropertyMock(return_value=False))
+    @patch("ops.framework.EventBase.defer")
+    def test_check_etcd_success(self, defer, *_):
```

Comment:
Needing 7 separate patches for a two-line unit test seems problematic. Is the test even that useful if we have to do this much tweaking before it can run?

Reply:
With these initial `test_charm.py` unit tests we are only interested in the expected charm behavior so that is why there are so many patches. Ideally in the future after refactoring we can simplify the mocks so there are far fewer after modifying the underlying methods. The goal of this initial unit test is to set a base expectation of how the charm operates before we start refactoring to ensure that as we refactor we do not alter charm functionality.

Reply:
As @dvdgomez mentioned, the real purpose here is to provide the unit tests that cover the existing codebase with the minimum amount of change. If there are refactor efforts needed to the code itself, then the unit tests exist already to express the intent of the methods and that things work. The refactoring would come in a subsequent patch in my mind.